Kernel/FileSystem/Ext2Base.h

```diff
@@ -0,0 +1,307 @@
+    u16 group_descriptor_size_num; // Group desc. size if INCOMPAT_64BIT is set
+    u32 default_mount_options;
+    u32 first_meta_block_group;   // First metablock group
+    u32 fs_create_time;           // When the filesystem was created
+    u32 journal_inode_backup[17]; // Backup of the journal inode
+    u32 num_blocks_high;          // Blocks count high 32bits
+    u32 num_reserved_blocks_high; // Reserved blocks count high 32 bits
+    u32 num_free_blocks_high;     // Free blocks count
+    u16 min_extra_inode_size;     // All inodes have at least # bytes
+    u16 want_extra_inode_size;    // New inodes should reserve # bytes
+    u32 misc_flags;               // Miscellaneous flags
+    u16 raid_stride;              // RAID stride
+    u16 mmp_interval;             // # seconds to wait in MMP checking
+    u64 mmp_block;                // Block for multi-mount protection
+    u32 raid_stripe_width;        // blocks on all data disks (N*stride)
+    u8 log_groups_per_flex;       // FLEX_BG group size
+    u8 reserved_char_pad;
+    u16 reserved_pad;  // Padding to next 32bits
+    u32 reserved[162]; // Padding to the end of the block
+
+    InodeIndex first_inode() const
+    {
+        if (rev_level == to_underlying(Revision::Original))
+            return to_underlying(Ext2::InodeNums::FirstInode);
+        else
+            return first_inode_num;
+    }
+    size_t inode_size() const
+    {
+        if (rev_level == to_underlying(Revision::Original))
+            return good_old_inode_size;
+        else
+            return inode_size_num;
+    }
+    size_t block_size() const
+    {
+        return (1 << min_block_log_size) << block_log_size;
+    }
+    size_t fragment_size() const
+    {
+        return (1 << min_fragment_log_size) << fragment_log_size;
+    }
+    // Returns the exponent of the block's size.
+    size_t block_size_bits() const
+    {
+        return block_log_size + min_block_log_size;
+    }
+    size_t inodes_per_block() const
+    {
+        return block_size() / inode_size();
+    }
+    size_t fragments_per_block() const
+    {
+        return block_size() / fragment_size();
+    }
+    size_t addresses_per_block() const
+    {
+        return block_size() / sizeof(u32);
+    }
+    size_t group_descriptor_size() const
+    {
+        if (incompatible_features & ext4_feature_incompat_64bit)
+            return group_descriptor_size_num;
+        else
+            return min_group_descriptor_size;
+    }
+    size_t group_descriptors_per_block() const
+    {
+        return block_size() / group_descriptor_size();
+    }
+};
+
+// The new version of the directory entry.  Since EXT2 structures are
+// stored in intel byte order, and the name_len field could never be
+// bigger than 255 chars, it's safe to reclaim the extra byte for the
+// file_type field.
+
+struct DirEntry {
+    u32 inode;   // Inode number
+    u16 rec_len; // Directory entry length
+    u8 name_len; // Name length
+    u8 file_type;
```

Comment:
Is this the file type as defined below? In that case, it should be a `FileType file_type`.

Reply:
OK, so if I were to change this to an `Ext2::FileType`, I'd have to change more files than I do right now (`FileSystem.cpp` and `FileSystem.h` also use `u8`s for the file type), is that OK?

Reply:
The generic file system should not require any changes as far as I can tell, you only have to modify `Ext2FSInode::traverse_as_directory` to pass on the enum as its underlying type.

Reply:
https://github.com/SerenityOS/serenity/blob/master/Kernel/FileSystem/FileSystem.h#L49 🤨